### ISSUE TITLE
CloudExceptions are missing Message

### ISSUE THREAD

**Issue by tgc123** [Azure/azure-sdk-for-java#1531]
Related to #1519, many operations seem to throw empty CloudExceptions.

The reason for the Exception is written into log, but CloudException still has null as message.

For example:
```
try {
    azure.virtualMachineScaleSets()
        .define("test1")
    	.withRegion("iceland")
    	.withExistingResourceGroup(group)
    	.withSku(VirtualMachineScaleSetSkuTypes.STANDARD_D1_V2)
    	.withExistingPrimaryNetworkSubnet(network, subnet)
    	.withoutPrimaryInternetFacingLoadBalancer()
    	.withoutPrimaryInternalLoadBalancer()
    	.withPopularLinuxImage(KnownLinuxVirtualMachineImage.UBUNTU_SERVER_16_04_LTS)
    	.withRootUsername("secret")
    	.withRootPassword("itsNOTsecret12345")
    	.withoutOverProvisioning()
    	.withCapacity(1)
    	.create();
} catch (CloudException ce) {
    System.err.println(ce + ": " + ce.getMessage());
}
```

The message in the end is:
`com.microsoft.azure.CloudException: null`

In the log entries, however I see a good description for the error:
```
[RxIoScheduler-2] INFO com.microsoft.azure.management.compute.VirtualMachineScaleSets createOrUpdate - 527-byte body:
{"error":{"code":"LocationNotAvailableForResourceType","message":"The provided location 'iceland' is not available for resource type 'Microsoft.Compute/virtualMachineScaleSets'. List of available regions for the resource type is 'eastus,eastus2,westus,centralus,northcentralus,southcentralus,northeurope,westeurope,eastasia,southeastasia,japaneast,japanwest,australiaeast,australiasoutheast,brazilsouth,southindia,centralindia,westindia,canadacentral,canadaeast,westus2,westcentralus,uksouth,ukwest,koreacentral,koreasouth'."}}
```

CloudException also contains field body (CloudError). The body should contain errorcode, message and details. In this case all of them were empty.

**Comment by anuchandy** [comment]
@tgc123 can you share the complete stack trace?

**Comment by tgc123** [comment]
Here's the trace. It seems that Jackson doesn't deserialize the bodyString for some reason and it returns an empty CloudError.

`com.microsoft.azure.AzureClient.createExceptionFromResponse(AzureClient.java:591), com.microsoft.azure.AzureClient.access$000(AzureClient.java:34), com.microsoft.azure.AzureClient$1.call(AzureClient.java:135), com.microsoft.azure.AzureClient$1.call(AzureClient.java:132), rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:69), retrofit2.adapter.rxjava.RxJavaCallAdapterFactory$RequestArbiter.request(RxJavaCallAdapterFactory.java:173), rx.Subscriber.setProducer(Subscriber.java:211), rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:102), retrofit2.adapter.rxjava.RxJavaCallAdapterFactory$CallOnSubscribe.call(RxJavaCallAdapterFactory.java:152), retrofit2.adapter.rxjava.RxJavaCallAdapterFactory$CallOnSubscribe.call(RxJavaCallAdapterFactory.java:138), rx.Observable.unsafeSubscribe(Observable.java:10142), rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48), rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33), rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48), rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30), rx.Observable.unsafeSubscribe(Observable.java:10142), rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48), rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33), rx.Observable.unsafeSubscribe(Observable.java:10142), rx.internal.util.ScalarSynchronousObservable$3.call(ScalarSynchronousObservable.java:235), rx.internal.util.ScalarSynchronousObservable$3.call(ScalarSynchronousObservable.java:228), rx.Observable.unsafeSubscribe(Observable.java:10142), rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48), rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33), rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48), rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30), rx.Observable.unsafeSubscribe(Observable.java:10142), rx.internal.operators.OperatorSubscribeOn$1.call(OperatorSubscribeOn.java:94), rx.internal.schedulers.CachedThreadScheduler$EventLoopWorker$1.call(CachedThreadScheduler.java:230), rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55), java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511), java.util.concurrent.FutureTask.run(FutureTask.java:266), java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180), java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293), java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142), java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617), java.lang.Thread.run(Thread.java:745)
`

Line 588 in AzureClient still contains the bodyString:
`{"error":{"code":"LocationNotAvailableForResourceType","message":"The provided location 'iceland' is not available for resource type 'Microsoft.Compute/virtualMachineScaleSets'. List of available regions for the resource type is 'eastus,eastus2,westus,centralus,northcentralus,southcentralus,northeurope,westeurope,eastasia,southeastasia,japaneast,japanwest,australiaeast,australiasoutheast,brazilsouth,southindia,centralindia,westindia,canadacentral,canadaeast,westus2,westcentralus,uksouth,ukwest,koreacentral,koreasouth'."}}`

But after line 589 (deserialize), the CloudError appears empty.

However, if I modify bodyString and remove "error" level to make the bodyString look like this:
`{"code":"LocationNotAvailableForResourceType","message":"The provided location 'iceland' is not available for resource type 'Microsoft.Compute/virtualMachineScaleSets'. List of available regions for the resource type is 'eastus,eastus2,westus,centralus,northcentralus,southcentralus,northeurope,westeurope,eastasia,southeastasia,japaneast,japanwest,australiaeast,australiasoutheast,brazilsouth,southindia,centralindia,westindia,canadacentral,canadaeast,westus2,westcentralus,uksouth,ukwest,koreacentral,koreasouth'."}`

Then the deserialize works ok and CloudError gets populated:
`com.microsoft.azure.CloudException: The provided location 'iceland' is not available for resource type 'Microsoft.Compute/virtualMachineScaleSets'. List of available regions for the resource type is 'eastus,eastus2,westus,centralus,northcentralus,southcentralus,northeurope,westeurope,eastasia,southeastasia,japaneast,japanwest,australiaeast,australiasoutheast,brazilsouth,southindia,centralindia,westindia,canadacentral,canadaeast,westus2,westcentralus,uksouth,ukwest,koreacentral,koreasouth'.`

**Comment by nwarriorch** [comment]
I hit the same bug today. with the following code on beta 5. 

azure.virtualMachines()
                .define("vm-6795842")
                .withRegion(Region.US_WEST2)
                .withNewResourceGroup("myresourcegroup")
                .withNewPrimaryNetwork("10.0.0.0/16")
                .withPrimaryPrivateIpAddressDynamic()
                .withoutPrimaryPublicIpAddress()
                .withLatestLinuxImage("OpenLogic", "CentOS", "6.6")
                .withRootUsername("user")
                .withRootPassword("password")
                .withUnmanagedDisks()
                .withSize(VirtualMachineSizeTypes.STANDARD_A0)
                .create();

Stack trace is partial. 

at com.microsoft.azure.AzureClient.createExceptionFromResponse(AzureClient.java:591)
	at com.microsoft.azure.AzureClient.access$000(AzureClient.java:34)
	at com.microsoft.azure.AzureClient$1.call(AzureClient.java:135)
	at com.microsoft.azure.AzureClient$1.call(AzureClient.java:132)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:69)
	at retrofit2.adapter.rxjava.RxJavaCallAdapterFactory$RequestArbiter.request(RxJavaCallAdapterFactory.java:173)
	at rx.Subscriber.setProducer(Subscriber.java:211)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:102)
	at retrofit2.adapter.rxjava.RxJavaCallAdapterFactory$CallOnSubscribe.call(RxJavaCallAdapterFactory.java:152)
	at retrofit2.adapter.rxjava.RxJavaCallAdapterFactory$CallOnSubscribe.call(RxJavaCallAdapterFactory.java:138)
	at rx.Observable.unsafeSubscribe(Observable.java:10142)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.Observable.unsafeSubscribe(Observable.java:10142)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.Observable.unsafeSubscribe(Observable.java:10142)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.Observable.unsafeSubscribe(Observable.java:10142)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:248)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:148)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:77)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:77)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:77)
	at rx.internal.producers.SingleProducer.request(SingleProducer.java:65)
	at rx.Subscriber.setProducer(Subscriber.java:211)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:102)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:102)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:102)
	at rx.internal.operators.OperatorSingle$ParentSubscriber.onCompleted(OperatorSingle.java:110)
	at rx.internal.operators.DeferredScalarSubscriber.complete(DeferredScalarSubscriber.java:102)
	at rx.internal.operators.OnSubscribeTakeLastOne$TakeLastOneSubscriber.onCompleted(OnSubscribeTakeLastOne.java:57)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onCompleted(OnSubscribeMap.java:97)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.emitLoop(OperatorMerge.java:656)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.emit(OperatorMerge.java:568)
	at rx.internal.operators.OperatorMerge$InnerSubscriber.onCompleted(OperatorMerge.java:857)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.emitLoop(OperatorMerge.java:656)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.emit(OperatorMerge.java:568)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onCompleted(OperatorMerge.java:281)
	at rx.internal.operators.OperatorMapNotification$MapNotificationSubscriber.tryEmit(OperatorMapNotification.java:155)
	at rx.internal.operators.OperatorMapNotification$MapNotificationSubscriber.onCompleted(OperatorMapNotification.java:121)
	at rx.internal.operators.OnSubscribeDoOnEach$DoOnEachSubscriber.onCompleted(OnSubscribeDoOnEach.java:70)
	at rx.internal.operators.OperatorSubscribeOn$1$1.onCompleted(OperatorSubscribeOn.java:68)
	at rx.internal.operators.OnSubscribeDoOnEach$DoOnEachSubscriber.onCompleted(OnSubscribeDoOnEach.java:70)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onCompleted(OnSubscribeMap.java:97)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.emitLoop(OperatorMerge.java:656)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.emit(OperatorMerge.java:568)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onCompleted(OperatorMerge.java:281)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onCompleted(OnSubscribeMap.java:97)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onCompleted(OnSubscribeMap.java:97)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.emitLoop(OperatorMerge.java:656)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.emit(OperatorMerge.java:568)
	at rx.internal.operators.OperatorMerge$InnerSubscriber.onCompleted(OperatorMerge.java:857)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onCompleted(OnSubscribeMap.java:97)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.emitLoop(OperatorMerge.java:656)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.emit(OperatorMerge.java:568)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onCompleted(OperatorMerge.java:281)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onCompleted(OnSubscribeMap.java:97)
	at rx.internal.producers.SingleProducer.request(SingleProducer.java:75)
	at rx.Subscriber.setProducer(Subscriber.java:211)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:102)
	at rx.internal.operators.OperatorSingle$ParentSubscriber.onCompleted(OperatorSingle.java:110)
	at rx.internal.operators.OperatorTake$1.onNext(OperatorTake.java:80)
	at rx.internal.operators.OnSubscribeFilter$FilterSubscriber.onNext(OnSubscribeFilter.java:76)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.emitScalar(OperatorMerge.java:395)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.tryEmit(OperatorMerge.java:355)
	at rx.internal.operators.OperatorMerge$InnerSubscriber.onNext(OperatorMerge.java:846)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.emitScalar(OperatorMerge.java:511)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.tryEmit(OperatorMerge.java:466)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:244)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:148)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:77)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.emitScalar(OperatorMerge.java:511)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.tryEmit(OperatorMerge.java:466)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:244)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:148)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:77)
	at retrofit2.adapter.rxjava.RxJavaCallAdapterFactory$RequestArbiter.request(RxJavaCallAdapterFactory.java:173)
	at rx.Subscriber.setProducer(Subscriber.java:211)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.setProducer(OnSubscribeMap.java:102)
	at retrofit2.adapter.rxjava.RxJavaCallAdapterFactory$CallOnSubscribe.call(RxJavaCallAdapterFactory.java:152)
	at retrofit2.adapter.rxjava.RxJavaCallAdapterFactory$CallOnSubscribe.call(RxJavaCallAdapterFactory.java:138)
	at rx.Observable.unsafeSubscribe(Observable.java:10142)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.Observable.unsafeSubscribe(Observable.java:10142)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.Observable.unsafeSubscribe(Observable.java:10142)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:248)
	at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:148)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:77)
	at rx.internal.operators.OnSubscribeRedo$2$1.onNext(OnSubscribeRedo.java:244)
	at rx.internal.util.ScalarSynchronousObservable$ScalarAsyncProducer.call(ScalarSynchronousObservable.java:200)
	at rx.internal.util.ScalarSynchronousObservable$2$1.call(ScalarSynchronousObservable.java:114)
	at rx.internal.schedulers.CachedThreadScheduler$EventLoopWorker$1.call(CachedThreadScheduler.java:230)
	at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

**Comment by TimoKramer** [comment]
I've got the same problem: no message and no hint on what is wrong.

**Comment by jianghaolu** [comment]
This is fixed in 1.0.0.

The cause is in beta5, the default constructor of `RestClient.Builder` takes `JacksonAdapter` which doesn't handle the Azure specifics as `AzureJacksonAdapter` does, including handling the error body response. (https://github.com/Azure/autorest-clientruntime-for-java/blob/v1.0.0-beta5/client-runtime/src/main/java/com/microsoft/rest/RestClient.java#L210)

1.0.0 is coming out latest next Monday. To work around, please create your own `RestClient` with `.withSerializerAdapter(new AzureJacksonAdapter()).withResponseBuilderFactory(new AzureResponseBuilder.Factory())`.
